### PR TITLE
Site Settings: Breakpoints - Added a validator for the breakpoint controls (ED-2605)

### PR DESCRIPTION
**Pull request by udidol**
Site Settings: Breakpoints - Added a validator for the breakpoint controls, which makes sure selected breakpoint values remain within the bounds of neighboring breakpoints' min/max limits. For example: If a mobile's max value is 600px, the validator makes sure that if a user edits the tablet breakpoint, and tries to give it a value of 599px, the value will automatically reset itself to the previous selected value which was above 600px. Same goes for upper-limit breakpoints.

In addition, a total minimum of 320px and maximum of 5120px was added into the validation conditions.

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

* Site Settings: Breakpoints - Added a validator for the breakpoint controls to prevent invalid values.

## Quality assurance

- [X] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: assets/dev/js/editor/components/validator/breakpoint.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+const NumberValidator = require( 'elementor-validator/number' );
+
+module.exports = NumberValidator.extend( {
```
Comment (kobizz) [inline assets/dev/js/editor/components/validator/breakpoint.js:3]: Write it as an ES6 class

File: assets/dev/js/editor/controls/base-data.js [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
 	TagsBehavior = require( 'elementor-dynamic-tags/control-behavior' ),
 	Validator = require( 'elementor-validator/base' ),
+	NumberValidator = require( 'elementor-validator/number' ),
+	BreakpointValidator = require( 'elementor-validator/breakpoint' ),
 	ControlBaseDataView;
 
 ControlBaseDataView = ControlBaseView.extend( {
+	validatorTypes: {
+		base: Validator,
+		number: NumberValidator,
```
Comment (kobizz) [inline assets/dev/js/editor/controls/base-data.js:11]: Use first letter upper case since they are class names

Comment (kobizz) [inline assets/dev/js/editor/controls/base-data.js:11]: Are they in use?

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/components/validator/breakpoint.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	breakpointKeysArray: [],
+
+	getDefaultSettings: function() {
+		return {
+			validationTerms: {
+				// Min width we allow for mobile
+				min: 320,
```
Comment (kobizz) [inline assets/dev/js/editor/components/validator/breakpoint.js:10]: You don't need the min definition since we added min value for mobile

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/components/validator/breakpoint.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	},
+
+	validateMinMaxForBreakpoint: function( newValue, validationTerms ) {
+		let isValid = true;
+
+		if ( ! this.breakpointKeysArray.length ) {
+			this.breakpointKeysArray = Object.keys( elementorFrontend.config.responsive.activeBreakpoints );
+		}
+
+		const breakpointIndex = this.breakpointKeysArray.indexOf( validationTerms.breakpointName ),
```
Comment (kobizz) [inline assets/dev/js/editor/components/validator/breakpoint.js:37]: Use `Stylesheet.getDeviceMinBreakpoint( deviceName );` for getting the min breakpoint

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/base-data.js (hunk 2 above)
Comment (udidol) [reply assets/dev/js/editor/controls/base-data.js:11]: Not currently but this is generic so any control can add any validator

Comment (udidol) [reply assets/dev/js/editor/controls/base-data.js:11]: ?

File: assets/dev/js/editor/components/validator/breakpoint.js (hunk 4 above)
Comment (udidol) [reply assets/dev/js/editor/components/validator/breakpoint.js:37]: Done.

File: assets/dev/js/editor/components/validator/breakpoint.js (hunk 3 above)
Comment (udidol) [reply assets/dev/js/editor/components/validator/breakpoint.js:10]: Done.

File: assets/dev/js/editor/controls/base-data.js (hunk 2 above)
Comment (udidol) [reply assets/dev/js/editor/controls/base-data.js:12]: Done.

Comment (udidol) [reply assets/dev/js/editor/controls/base-data.js:12]: Done.

File: assets/dev/js/editor/components/validator/breakpoint.js (hunk 1 above)
Comment (udidol) [reply assets/dev/js/editor/components/validator/breakpoint.js]: Done.

File: assets/dev/js/editor/controls/base-data.js [hunk 5]
```diff
@@ -1,9 +1,18 @@
...
 var ControlBaseView = require( 'elementor-controls/base' ),
 	TagsBehavior = require( 'elementor-dynamic-tags/control-behavior' ),
 	Validator = require( 'elementor-validator/base' ),
+	NumberValidator = require( 'elementor-validator/number' ),
 	ControlBaseDataView;
 
 ControlBaseDataView = ControlBaseView.extend( {
+	validatorTypes: {
+		Base: Validator,
+		Number: NumberValidator,
```
Comment (kobizz) [inline assets/dev/js/editor/controls/base-data.js:12]: Use these references

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/components/validator/breakpoint.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+		// If there is a breakpoint below the currently edited breakpoint, check that the value is not under the bottom
+		// breakpoint's value.
+		if ( bottomBreakpoint && ( newValue <= bottomBreakpoint ) ) {
+			isValid = false;
+		}
+
+		// If there is a breakpoint above the currently edited breakpoint, check that the value is not above the top
+		// breakpoint's value.
+		if ( topBreakpointKey && newValue >= elementorFrontend.config.responsive.activeBreakpoints[ topBreakpointKey ].value ) {
```
Comment (kobizz) [inline assets/dev/js/editor/components/validator/breakpoint.js:53]: Check only for greater than

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/components/validator/breakpoint.js (hunk 6 above)
Comment (udidol) [reply assets/dev/js/editor/components/validator/breakpoint.js]: If I only check for a greater value, the validation will pass if a user puts the same value for mobile and tablet... which shouldn't happen.

**Review by kobizz — APPROVED** [review]

**Pull request merged**